Launching ydoc-server together with language-server

JaroslavTulach opened the pull request:
### Pull Request Description

- Another attempt to fix #9862 
- @Frizi would like to see `ydoc-server` running in the `enso` process
   - #10627 had to disable it as it was _giving an impression of a slower startup_
   - but now _the startup of Enso_ is no longer a problem with _native image_ executable
- thus let's include `ydoc-server` again in the `enso` executable 
   - but without increasing `enso` image size
   - e.g. let's use #13172 - but not for libraries, but for loading _Graal.js_ and _websocket emulation_
- by running `ydoc-server` in the same process we are one step closer to remove Akka networking layer

### Checklist & TODO

Inception [review has happened](https://docs.google.com/document/d/1u9C5ukgV76oQA4y8STXT9ZqbQoZJFbAIEf5jWq7B26w) on May 28, 2025 and based on that following action items (related to this PR) were identified:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- [x] Unit tests have been written where possible.
- [x] (Almost) no increase in `enso` executable size - [done](https://github.com/enso-org/enso/pull/13178#issuecomment-3369801447)
- [x] Measure startup slow down 
   - [done](https://github.com/enso-org/enso/pull/13178/files#r2109805132) - far less than **500ms** when built into _native image_
   - with the current _dual JVM mode_ the startup of the Ydoc server is slower - still less than 2s
   - the idea however is to bring it back below 0.5s by compiling the `ydoc-server` project into _"dual native library"_
- [x] Drop `ydoc-server-nodejs` source code - [done](https://github.com/enso-org/enso/pull/13178#discussion_r2113576580)
- [x] Drop `ydoc-server-nodejs` docker image - [done](https://github.com/enso-org/enso/pull/13178/commits/40e3ab54254eb477927cec661992cd3cfd717828) - there still is `ydoc-server-polyglot` docker image, in case we need 1:1 replacement for the docker image
- [x] Add `--ydoc-port  <number>` option to `language-server` - **kinda done** by https://github.com/enso-org/enso/pull/13178#discussion_r2415882922
- [x] ~Expand [project/open](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-project-manager.md#projectopen) to provide also `ydoc-port` value~
- [x] Change `corepack pnpm dev:gui` to use this GraalVM based Ydoc server: [done](https://github.com/enso-org/enso/pull/13178/files#r2109892028)
- [x] Ensure `.AppImage` & co. are OK using the embedded `ydoc-server` - works since 76a0265
- [x] Ensure cloud is OK using the embedded `ydoc-server`

JaroslavTulach committed:
Launching ydoc-server together with language-server

JaroslavTulach commented on lib/java/ydoc-server/src/main/java/org/enso/ydoc/server/YdocServerImpl.java:
```diff
@@ -0,0 +1,29 @@
+      var ydoc = Ydoc.builder().hostname(hostname).port(port).build();
+      ydoc.start();
+      var now = System.currentTimeMillis();
+      log.warn("Ydoc server at {}:{} started in {} ms", hostname, port, now - then);
```
With three subsequent start ups I see:
```
Ydoc server at localhost:1234 started in 452 ms
Ydoc server at localhost:1234 started in 347 ms
Ydoc server at localhost:1234 started in 327 ms
```
e.g. we can assume that the slowdown is at most half a second. That sounds more acceptable than 5s in the HotSpot JVM mode. Let the integration continue...

JaroslavTulach commented:
Snapshot of memory usage of the JavaScript inside of `org.enso.runner.Main` process:

![2MB](https://github.com/user-attachments/assets/16ceac86-52e1-4db5-b54a-b1ba72c51256)

The Y.doc JavaScript data structures are estimated to fit into 2MB of memory.

JaroslavTulach committed:
Removing start of node.js based Y.doc server. Using GraalVM based one at the same port.

JaroslavTulach commented on app/electron-client/src/server.ts:
```diff
@@ -148,42 +147,7 @@ export class Server {
               reject(err)
             }
-            const server = httpsServer ?? httpServer
-            if (!IS_ELECTRON_DEV_MODE) {
```
- 57436c8c9442ee4f940b621aa4ef2e0b39d06341 removes start of the node.js based Y.doc server
- when doing `corepack pnpm dev:gui`
- together with `sbt runProjectManagerDistribution` the IDE still continues to work
- do you want to remove more code, @Frizi? Feel free to commit directly to this PR.

JaroslavTulach replied:
Removing whole `ydoc-server-nodejs` project in 7033120da4c8999fc5a8a24a380d30029c1a72a7

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/YdocInBackend9862

JaroslavTulach committed:
Let the default Ydoc port be 5976

JaroslavTulach committed:
No need for ydoc-server-nodejs project

JaroslavTulach committed:
Note in changelog

JaroslavTulach committed:
Removing docker image build support for ydoc-server-nodejs

JaroslavTulach commented:
- there is a significant problem:
   - CI is running [out of memory on Linux](https://github.com/enso-org/enso/actions/runs/15324272689/job/43114773753?pr=13178#step:10:4382) when building the `enso` _native image_
   - CI [Windows run is fine](https://github.com/enso-org/enso/actions/runs/15324272689/job/43114774256?pr=13178#step:10:3074) - it just fails on the _native image_ post mortem check - it is too big - e.g. **606MB** - but we could increase the limit...
   - CI [Mac run also fails](https://github.com/enso-org/enso/actions/runs/15324272689/job/43114775183?pr=13178#step:10:5558) on GC related problems
- I am afraid we have to suspend this PR until we make the _native image_ smaller
- for example by #13172, ach.

#### Update on Oct 6, 2025

- now this PR actually uses the #13172 for running itself
- e.g. the [websocket emulation & Graal.js interpreter](https://github.com/enso-org/enso/pull/13178#issuecomment-3369801447) are executed in _dual JVM_ 
- without any impact on the size of the _native image_

JaroslavTulach committed:
Merged with most recent develop

JaroslavTulach committed:
Launch the the internal YdocServer in the other JVM

JaroslavTulach committed:
Splitting Y.doc server implementation and registration

JaroslavTulach commented:
- after recent changes we can build `enso` native image on
   - Linux - [569MB](https://github.com/enso-org/enso/actions/runs/18263042619/job/52007865317?pr=13178#step:10:3429) which is almost identical to [568MB](https://github.com/enso-org/enso/actions/runs/18238997474/job/52008617506#step:10:2930) or today's `develop` branch
   - Windows - [544MB](https://github.com/enso-org/enso/actions/runs/18263042619/job/52007865575?pr=13178#step:10:3189) which is almost identical to [543MB](https://github.com/enso-org/enso/actions/runs/18238997474/job/52008617493#step:10:2985) of today's `develop` branch
- _image size wise_ we are OK to integrate

JaroslavTulach committed:
Reorganizing changelog

JaroslavTulach committed:
org.enso.ydoc.server module should be loaded dynamically by ydoc-server-registration

JaroslavTulach committed:
Unify two main classes into one

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/YdocInBackend9862

JaroslavTulach committed:
close operation on the OtherJvmClassLoader to match with hosted URLClassLoader.close()

JaroslavTulach commented:
- there is now [ydoc-server/Test failure](https://github.com/enso-org/enso/actions/runs/18272049079/job/52017227144?pr=13178#step:14:6707)

JaroslavTulach committed:
ydoc-server-nodejs is gone

JaroslavTulach committed:
Removing unused import adviced by prettier

JaroslavTulach committed:
Removing unused import

JaroslavTulach committed:
Aligning with 22f9edd76f94e014ad6dbd1d7a05460cfbeac5d7 and making the default Ydoc port 5976

JaroslavTulach commented on tools/ci/docker/engine/Dockerfile:
```diff
@@ -5,7 +5,7 @@ USER root
 ARG RPC_PORT=30001
 ARG DATA_PORT=30002
-ARG YDOC_PORT=1234
+ARG YDOC_PORT=5976
```
@PabloBuchu requested the change of port to 5976 to align with 22f9edd

JaroslavTulach replied:
Also had to change the default port in the electron app: https://github.com/enso-org/enso/pull/13178/commits/76a0265e61de0517d3f2a088f943ecac8763350b#r2415882922

JaroslavTulach committed:
Removing unused variables

JaroslavTulach commented:
- we have a [bazel problem](https://github.com/enso-org/enso/actions/runs/18301570053/job/52110458991?pr=13178#step:4:767)
- `sbt` is trying to build `.js` file and reports it `Cannot run program "corepack"`
- ideas and help welcomed @vitvakatu, @Akirathan

PabloBuchu commented on app/ide-desktop/client/src/server.ts:
```diff
@@ -87,7 +86,7 @@ const HTTP_STATUS_OK = 200
 const HTTP_STATUS_NOT_FOUND = 404
 const HTTP_STATUS_INTERNAL_SERVER_ERROR = 500
-const IS_ELECTRON_DEV_MODE = process.env.ELECTRON_DEV_MODE === 'true'
+// const IS_ELECTRON_DEV_MODE = process.env.ELECTRON_DEV_MODE === 'true'
```
commented code left

JaroslavTulach replied:
It is gone now.

PabloBuchu commented:
LGTM. After adjusting exposed port in Dockerfile all should be fixed. Lemme know when its merged and done in nightly so I can test in the cloud

vitvakatu commented on app/ide-desktop/client/src/server.ts:
```diff
@@ -251,48 +250,13 @@ export class Server {
-                console.log(`Ydoc server listening on localhost:5976`)
-              })
-            }
+            logger.warn('YDocs server is not run, new GUI may not work properly!')
```
I don’t understand the purpose of this message. We are not running it deliberately, then why warn the user?

JaroslavTulach replied:
Removed in most recent merge: 63690c1cee

vitvakatu commented on build.sbt:
```diff
@@ -533,6 +534,8 @@ lazy val componentModulesPaths =
     (`zio-wrapper` / Compile / exportedModuleBin).value,
     (`language-server-deps-wrapper` / Compile / exportedModuleBin).value,
     (`ydoc-polyfill` / Compile / exportedModuleBin).value,
+    (`ydoc-server` / Compile / exportedModuleBin).value,
```
Because now ydoc-server is a dependency of `buildEngineDistribution`, it started failing bazel build. The correct fix is: 
- generate ydoc-server using bazel
- provide it for sbt as artifact
- remove Ydoc.scala

Unfortunately, I’m not versed enough with sbt, @Akirathan should be able to help.

JaroslavTulach replied:
> * generate ydoc-server using bazel

How that would be done?

Akirathan replied:
I will try to implement a fix tomorrow. I don't think you need to generate `ydoc-server` with bazel now. Instead, ensure that the `.js` files are built using bazel, and then provide it to sbt via [environment variables](https://github.com/enso-org/enso/blob/6fa8d5f0b8d32a367ac65315f94a8bb884816a8e/project/BazelSupport.scala#L18-L23).

JaroslavTulach replied:
> the `.js` files are built using bazel, and then provide it to sbt via [environment variables](https://github.com/enso-org/enso/blob/6fa8d5f0b8d32a367ac65315f94a8bb884816a8e/project/BazelSupport.scala#L18-L23).

Of course! I didn't even notice anyone else would suggest otherwise.

JaroslavTulach replied:
Thanks Pavel for 7df1377

JaroslavTulach committed:
Merging with latest develop

JaroslavTulach committed:
Default Ydoc port is 5976

JaroslavTulach commented on app/gui/src/dashboard/services/LocalBackend.ts:
```diff
@@ -375,7 +375,7 @@ export default class LocalBackend extends Backend {
         jsonAddress: ipWithSocketToAddress(cachedProject.languageServerJsonAddress),
         binaryAddress: ipWithSocketToAddress(cachedProject.languageServerBinaryAddress),
-        ydocAddress: null,
+        ydocAddress: backend.Address('ws://localhost:5976'),
```
- before this change I had to specify the Ydoc port for unpacked AppImage manually:
- `enso/dist/ide/linux-unpacked/enso --engine.ydocUrl=http://localhost:5976`
- the AppImage connects to internal Ydoc server automatically
- here is a [proof internal Ydoc server](https://github.com/enso-org/enso/actions/runs/18369253779/job/52331838971?pr=13178#step:15:382) (with WebSocket emulated via Helidon library) is used

JaroslavTulach replied:
- Reapplying this change again (after few merges): 2a1b74e0d8
- Now `./run build ide` and executing `.AppImage` opens local project without issues

farmaazon commented on app/gui/src/dashboard/services/LocalBackend.ts:
```diff
@@ -375,7 +375,7 @@ export default class LocalBackend extends Backend {
         jsonAddress: ipWithSocketToAddress(cachedProject.languageServerJsonAddress),
         binaryAddress: ipWithSocketToAddress(cachedProject.languageServerBinaryAddress),
-        ydocAddress: null,
+        ydocAddress: backend.Address('ws://localhost:5976'),
```
The part where we set default port in case of null is here: https://github.com/enso-org/enso/blob/develop/app/gui/src/project-view/stores/project/index.ts#L68

It looks like something temporarily added (it's not even a constant). Your fix is better, as we should get information from Project Manager.

JaroslavTulach replied:
I guess I change the `1234` value to `5976` there as well... done in 7d06ee6

farmaazon reviewed (approved):
The JS part LGTM

JaroslavTulach committed:
The default port was changed to 5976

Commit:
bazel build file for ydoc-server-polyglot

Commit:
ydoc-server-polyglot/dist outputs just a single `main.cjs` file

Commit:
ydoc-server-polyglot/main.cjs is passed via system prop from Bazel to sbt

JaroslavTulach committed:
Can adding a comma change result of the whole Bazel build?

4e6 committed:
fix: ci nightly builds polyglot ydoc

Otherwise the release workflow will fail because the nodejs ydoc
package was removed.

4e6 commented:
@JaroslavTulach I pushed the commit that fixes nightly build and scheduled [nigthly release](https://github.com/enso-org/enso/actions/runs/18411503099) to verify.

4e6 reviewed (approved):
The nightly looks fine

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/YdocInBackend9862

JaroslavTulach committed:
Merge branch 'wip/jtulach/YdocInBackend9862' of enso:enso-org/enso into wip/jtulach/YdocInBackend9862

hubertp commented on lib/java/jvm-interop/src/main/java/org/enso/jvm/interop/api/OtherJvmClassLoader.java:
```diff
@@ -65,7 +72,7 @@ boolean isMemberReadable(String member) {
   @ExportMessage
   boolean isMemberInvocable(String member) {
-    return "addPath".equals(member);
+    return "addPath".equals(member) || "findLibraries".equals(member) || "close".equals(member);
```
Wouldn't it be just easier to read with a list and check with `contains`?

JaroslavTulach replied:
- consistency among various `InteropLibrary` messages is subject to may _innovation_
- alas usually it is subject to clueless evolution (like in this case)
- it starts simple `return "addPath".equals(member);` and then it gets more and more convoluted
- you'd like a change to
- https://github.com/enso-org/enso/blob/a20924def841c298553920ac6c6cb75879e9dfcb/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java#L215
- yeah, that may be better
- anyway consistency among various implementations spread around the code base would remain a problem

hubertp commented on lib/java/ydoc-server-registration/src/main/java/module-info.java:
```diff
@@ -0,0 +1,13 @@
+  requires org.enso.jvm.interop;
+  requires static org.graalvm.nativeimage;
+
+  // only register service, otherwise the module isn't need
```
```suggestion
  // only register service, otherwise the module isn't needed
```

hubertp commented on lib/java/ydoc-server-registration/src/main/java/org/enso/ydoc/server/registration/YdocServerImpl.java:
```diff
@@ -0,0 +1,53 @@
+                  .getCodeSource()
+                  .getLocation()
+                  .toURI());
+      var ydocServerJar = new File(myJar.getParentFile(), "ydoc-server.jar");
```
`ydoc-server.jar` ideally that would be provided externally. Someone changes the name of the artifact in sbt/Bazel and then we will end up chasing a bug.

JaroslavTulach replied:
> ideally would be provided externally

I am open to patches, but otherwise I can only say: If the name changes, then many things go wrong:

<img width="408" height="1099" alt="image" src="https://github.com/user-attachments/assets/438b2209-c081-4c22-b97e-2648b98b85cf" />

hubertp commented on lib/java/ydoc-server/src/main/java/org/enso/ydoc/server/Main.java:
```diff
@@ -16,21 +21,40 @@ public static void main(String[] args) throws Exception {
+
+      var now = System.currentTimeMillis();
+      var took = now - then;
+      log.warn("Ydoc server at {}:{} started in {} ms", hostname, port, took);
```
```suggestion
      log.debug("Ydoc server at {}:{} started in {} ms", hostname, port, took);
```

Warnings are always seen by users. There is no need for that here.

hubertp commented on lib/java/ydoc-server/src/main/java/org/enso/ydoc/server/Main.java:
```diff
@@ -16,21 +21,40 @@ public static void main(String[] args) throws Exception {
+      var now = System.currentTimeMillis();
+      var took = now - then;
+      log.warn("Ydoc server at {}:{} started in {} ms", hostname, port, took);
+      System.err.printf("Ydoc server at %s:%s started in %d ms\n", hostname, port, took);
```
```suggestion
```

We definitely don't want to print that stderr

hubertp commented on lib/java/ydoc-server/src/test/java/org/enso/ydoc/server/YdocTest.java:
```diff
@@ -37,8 +37,8 @@
   private static final int WEB_SERVER_PORT = 44556;
-  private static final String YDOC_URL = "ws://localhost:1234/project/";
-  private static final String HEALTHCHECK_URL = "http://localhost:1234/_health";
+  private static final String YDOC_URL = "ws://localhost:5976/project/";
+  private static final String HEALTHCHECK_URL = "http://localhost:5976/_health";
```
Surely we could make use of the hardcoded `DEFAULT_PORT` variable, no?

JaroslavTulach replied:
First and foremost, the port shall not be hardcoded at all. Having a port like this in a desktop application prevents usage on multiuser systems - e.g. single server in a lab used by multiple students in a class - fixing that would fix the real problem. 

> use of the hardcoded DEFAULT_PORT variable

That's only putting a lipstick on a pig...

JaroslavTulach replied:
The original code was:
```
  private static final String YDOC_URL = "ws://localhost:1234/project/";
  private static final String HEALTHCHECK_URL = "http://localhost:1234/_health";
```
and my change is thus keeping the spirit of the original code. Leaving as is.

hubertp reviewed (approved):
I don't like that we have this hardcoded port copied over and over in multiple places. But it has been there before. 

But I do mind the extra logs/print err we seem to add when starting YDoc Server. Approving, but please consider the changes.

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/YdocInBackend9862

JaroslavTulach committed:
Removing System.err printf

Co-authored-by: Hubert P <hubert.plociniczak@gmail.com>

JaroslavTulach committed:
Using debug log level

Co-authored-by: Hubert P <hubert.plociniczak@gmail.com>

JaroslavTulach committed:
Fixing a typo

Co-authored-by: Hubert P <hubert.plociniczak@gmail.com>

JaroslavTulach commented:
- @4e6, `project-manager` tests are failing again https://github.com/enso-org/enso/actions/runs/18456787495/job/52588422347?pr=13178#step:14:6518
```
 INFO ide_ci::program::command: sbt ℹ️ [error]     org.enso.projectmanager.infrastructure.languageserver.LanguageServerGatewaySpec
 INFO ide_ci::program::command: sbt ℹ️ [error]     org.enso.projectmanager.protocol.ProjectShutdownSpec
 INFO ide_ci::program::command: sbt ℹ️ [error]     org.enso.projectmanager.protocol.ProjectManagementApiSpec
```
- Given `project-manager` has been removed in #14116
- Can we disable these tests in my branch?
- Failed again. Disabled in 6b45001

JaroslavTulach committed:
Disabling project-manager tests that are randomly failing while testing no longer used code

JaroslavTulach committed:
Merging with latest develop

JaroslavTulach committed:
Removing duplicated functions

JaroslavTulach committed:
Another merge with develop

JaroslavTulach committed:
Adding ydocAddress again

JaroslavTulach commented:
- [Package New IDE on Linux fails](https://github.com/enso-org/enso/actions/runs/18518944636/job/52775509502?pr=13178#step:15:257)
- [Package New IDE on Windows fails](https://github.com/enso-org/enso/actions/runs/18518944636/job/52777749715?pr=13178#step:15:219)
- both are executing `corepack pnpm -r --filter enso ide-integration-test`

```
 1) tests\localWorkflow.spec.ts:27:1 › Project Duplicate ──────────────────────────────────────────

    Error: Timed out 60000ms waiting for expect(locator).toHaveCount(expected)

    Locator: locator('.GraphNode')
    Expected: 1
    Received: 8
    Call log:
      - Expect "toHaveCount" with timeout 60000ms
      - waiting for locator('.GraphNode')
        11 × locator resolved to 0 elements
           - unexpected value "0"
        52 × locator resolved to 8 elements
           - unexpected value "8"


       at electronTest.ts:122

      120 |   await expect(newProjectTab).toBeVisible()
      121 |   await newProjectTab.click()
    > 122 |   await expect(page.locator('.GraphNode')).toHaveCount(1, { timeout: 60000 })
          |                                            ^
      123 |
      124 |   const tableViz = page.locator('.TableVisualization')
      125 |   await expect(tableViz).toBeVisible({ timeout: 30000 })
        at createNewProject (C:\runner\_work\enso\enso\app\electron-client\tests\electronTest.ts:122:44)
        at C:\runner\_work\enso\enso\app\electron-client\tests\localWorkflow.spec.ts:30:3

    Error Context: test-results\localWorkflow-Project-Duplicate\error-context.md
```

- my problem is that I cannot run the tests locally!
- CCing @otulach
- hopefully resolved by [clicking the New Project button](https://github.com/enso-org/enso/pull/13178/files#r2433038593)

vitvakatu commented on app/gui/.dev-env:
Please update your local .dev-env; it is currently being reverted in this PR.

JaroslavTulach replied:
Done in 18a806d

JaroslavTulach committed:
Returning app/gui/.dev-env back to original state

JaroslavTulach committed:
Click on New Project button, not 10, 100 co-ordinate

JaroslavTulach commented on app/electron-client/tests/localWorkflow.spec.ts:
```diff
@@ -102,7 +102,7 @@ test('Local Workflow', async ({ page, app, projectsDir }) => {
 
   // Leave function
-  await page.mouse.dblclick(10, 100)
+  await page.getByTestId('editor').getByText('New Project').dblclick()
```
- the [failing tests](https://github.com/enso-org/enso/pull/13178#issuecomment-3404888171) were caused by the "mis-click"
- as @Frizi puts it: _"The intention of that line was to exit the function by double-clicking the graph background. It probably hits a node."_
- I'd like to unblock myself and thus I'd rather simply click "New Project" in de779ed2e49d240007d81287e5a1b2f036faf8bf
- feel free to fine tune the double click later

JaroslavTulach commented:
Another [playwright failure](https://github.com/enso-org/enso/actions/runs/18534057380/job/52869236778?pr=13178#step:15:233) this time on Windows only:
```
Call log:
      - Expect "toHaveCount" with timeout 60000ms
      - waiting for locator('.GraphNode')
        10 × locator resolved to 0 elements
           - unexpected value "0"
        53 × locator resolved to 8 elements
           - unexpected value "8"


       at electronTest.ts:122

      120 |   await expect(newProjectTab).toBeVisible()
      121 |   await newProjectTab.click()
    > 122 |   await expect(page.locator('.GraphNode')).toHaveCount(1, { timeout: 60000 })
          |                                            ^
      123 |
      124 |   const tableViz = page.locator('.TableVisualization')
      125 |   await expect(tableViz).toBeVisible({ timeout: 30000 })
        at createNewProject (C:\runner\_work\enso\enso\app\electron-client\tests\electronTest.ts:122:44)
        at C:\runner\_work\enso\enso\app\electron-client\tests\localWorkflow.spec.ts:53:3
```

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/YdocInBackend9862

JaroslavTulach commented:
- @jdunkerley, @hubertp, @PabloBuchu looks like this PR may get green
- the problematic `playwright` tests on windows are [already green](https://github.com/enso-org/enso/actions/runs/18555376801/job/52894850380?pr=13178)
- can you please decide whether to integrate or not?
- the [IDEs are available](https://github.com/enso-org/enso/actions/runs/18555376801?pr=13178) for download

<img width="1180" height="1005" alt="image" src="https://github.com/user-attachments/assets/70ec572e-883c-4119-89ca-430391a9724b" />
